### DISCUSSION TITLE
Replicate file between different pool groups

### DISCUSSION THREAD

**Discussion by kofemann** [dCache/dcache#6332] [Ideas]
As of now, the **resilience manager**, as well as its predecessor **replica manager** operate only in a simple pool group. Though this makes sens in a general case, there are situation where we might consider to relax this limitation.

Imagine the following deployment: we want to replicate data to improve the resilience, however we want to be power efficient. To achieve that, a site deploys two type of servers - one high performance, for direct user IO, and a second one, low performance, but power effective, for replicas. Such setup will require two `links` and two `pool groups`:

```
psu create pgroup io-pools
psu create group replica-pools

psu create link user-io
psu set link user-io  -readpref=10 -writepref=10 -cachepref=0 -p2ppref=0
psu addto link user-io  io-pools

psu create link replicas 
psu set link replicas  -readpref=1 -writepref=1 -cachepref=0 -p2ppref=10
psu addto link replicas  replica-pools
```

The setup above makes read ad write more attractive from `io-pools` pool group. But in addition, we want to have a replica of the file tp be created in the `replica-pools` group.

## Alternatives

The dCache is powerful enough to achieve the desired functionality with the existing codebase. For example, the all pools in can be tagged with a tag corresponding to the hardware role, for example, `pool.tags=hw-class=IO` and `pool.tags=hw-class=resilience`. As **resilience manager** operates on a single pool group in addition to pool manager setup above a new pool group should be added that will include all pools from `io-pools` and `replica-pools`. The pool groups are not hierarchical, thus all polls ahave to be added on-by-one. Finally, the storage group replication rule have to explicitly restrict the replication into the same pool group:

```
psu create pgroup io-pools
psu create group replica-pools
psu create pgroup all-pools -resilient

# populate all pool groups
...

psu create link user-io
psu set link user-io  -readpref=10 -writepref=10 -cachepref=0 -p2ppref=0
psu addto link user-io  io-pools

psu create link replicas 
psu set link replicas  -readpref=1 -writepref=1 -cachepref=0 -p2ppref=10
psu addto link replicas  replica-pools

psu create unit -store foo:bar@osm
psu set storage unit foo:bar@osm -required=2 -onlyOneCopyPer=hw-class
```

**Comment by paulmillar** [comment]
OK.  If I've understood correctly, the desired functionality exists, once dCache is properly configured.

However, the configuration is a bit awkward.  In particular, adding a new pool is cumbersome, because the pool must be added to one of the hw-class poolgroups (either `io-pools` or `replica-pools`) and _also_ added to the poolgroup that resilience works with.

The desired innovation is that this be made simpler and easier to manage.

For example, one possible solution is to support having an auto-generated poolgroup from "all pools tagged `hw-class=IO`" (equiv. to `io-pools` above), an auto-generated poolgroup "all pools tagged `hw-class=replica`" (equiv. to `replica-pools` above), and an auto-generated poolgroup that is the union of the previous two poolgroups.

(That's just an illustrative example.)

With this model, tagging a pool is sufficient for it to be found (and used) be resilience.

**Reply by kofemann** [reply]
Actually, we already have dynamic pools groups based on tags. To make the setup simple, we need as well a possibility to build pool groups from groups:

```
psu create pgroup -dynamic -tags=hw-class=IO io-pools
psu create pgroup -dynamic -tags=hw-class=replica replica-pools
psu create pgroup all-pools
psu addto pgroup all-pools @io-pools @replica-pools
```

The symbol `@` will identify pool group, similar to netgroups in NIS
